src/katsdpmodels/models.py

```diff
@@ -0,0 +1,263 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Base functionality common to all model types."""
+
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+import io
+import urllib.parse
+import h5py
+from typing import Optional, Union, Any, ClassVar, Type, TypeVar, overload
+
+import numpy as np
+import strict_rfc3339
+
+
+_E = TypeVar('_E', bound='ModelError')
+_M = TypeVar('_M', bound='Model')
+_H = TypeVar('_H', bound='SimpleHDF5Model')
+
+
+class ModelError(ValueError):
+    """A model was found, but the content was incorrect."""
+
+    def __init__(self, *args) -> None:
+        super().__init__(*args)
+        self.url: Optional[str] = None
+        self.original_url: Optional[str] = None
+
+    @classmethod
+    def with_urls(cls: Type[_E], *args,
+                  url: Optional[str] = None, original_url: Optional[str] = None) -> _E:
+        exc = cls(args)
+        exc.url = url
+        exc.original_url = original_url
+        return exc
+
+
+class FileTypeError(ModelError):
+    """The file type (as determined by extension) was not recognised."""
+
+
+class ModelTypeError(ModelError):
+    """The ``model_type`` attribute was missing or did not match the expected value."""
+
+
+class ModelFormatError(ModelError):
+    """The ``model_format`` attribute was missing or did match a known value."""
+
+
+class DataError(ModelError):
+    """The model was missing some data or it had the wrong format."""
+
+
+class ChecksumError(DataError):
+    """The model did not match the checksum embedded in the filename."""
+
+
+class TooManyAliasesError(ModelError):
+    """The limit on the number of alias redirections was reached."""
+
+
+class Model(ABC):
+    """Base class for models.
+
+    Models can either be loaded from file-like objects or constructed directly.
+    Models loaded by the fetcher store the checksum of the raw data, and are
+    considered equal if the checksums match. Otherwise, equality is by object
+    identity.
+
+    Models loaded by the fetcher should not be modified, as they may be shared
+    by other users. Instead, make a copy and modify that.
+
+    Subclassing should generally be done in two layers:
+
+    1. A class that defines `model_type` and defines the interface for that
+       model type. This will be passed to :class:`.Fetcher` to indicate what
+       model type is expected. Due to limitations in mypy, this should not
+       use ``@abstractmethod`` for the interface methods.
+    2. A concrete implementation that defines `model_format`.
+    """
+
+    model_type: ClassVar[str]
+    model_format: ClassVar[str]
+    checksum: Optional[str] = None
+    target: Optional[str] = None
+    comment: Optional[str] = None
+    author: Optional[str] = None
+    created: Optional[datetime] = None
+
```

Comment:
How about version ? (this may be elsewhere - just commenting as I go)

Reply:
This wasn't in the design doc, so isn't in the implementation. Do you think it's worth trying to keep a version number AND a creation date?

Reply:
I must have missed it in the design doc, I do think a version is important, if nothing else to provide a more user friendly view of versioning (i.e. is this the next model on, or do I need to check the date range to see if I am missing something). Not a huge deal, but I think we would find uses for it.

Reply:
Can you add a comment to the design doc so that we can discuss it there with others?

Reply:
Will do.

Reply:
Done